app/jobs/generate_job.rb

```diff
@@ -3,61 +3,61 @@ class GenerateJob < ApplicationJob
   end
 
   before_enqueue do |job|
-    status.update(job_class: self.class)
     status.update(exam_name: job.arguments[0].name)
   end
 
   def perform(exam_template, num_copies, start)
     m_logger = MarkusLogger.instance
-    begin
-      progress.total = num_copies
-      template_path = File.join(
-        exam_template.base_path,
-        exam_template.filename
-      )
-      template_pdf = CombinePDF.load template_path
-      generated_pdf = CombinePDF.new
-      (start..start + num_copies - 1).each do |exam_num|
-        m_logger.log("Now generating: #{exam_num}")
-        pdf = Prawn::Document::new(margin: 15) do
-          exam_template.num_pages.times do |page_num|
-            qrcode_content = "#{exam_template.name}-#{exam_num}-#{page_num + 1}"
-            qrcode = RQRCode::QRCode.new qrcode_content, level: :l, size: 2
-            alignment = page_num % 2 == 0 ? :right : :left
-            render_qr_code(qrcode, align: alignment, dot: 4.0, stroke: false)
-            draw_text(qrcode_content,
-                 at: [alignment == :left ? 140 : bounds.width - 140 - qrcode_content.length * 6,
-                      bounds.height - 30])
-            start_new_page
-          end
+    progress.total = num_copies
+    template_path = File.join(
+      exam_template.base_path,
+      exam_template.filename
+    )
+    template_pdf = CombinePDF.load template_path
+    generated_pdf = CombinePDF.new
+    (start..start + num_copies - 1).each do |exam_num|
+      m_logger.log("Now generating: #{exam_num}")
+      pdf = Prawn::Document.new(margin: 15) do
+        exam_template.num_pages.times do |page_num|
+          qrcode_content = "#{exam_template.name}-#{exam_num}-#{page_num + 1}"
+          qrcode = RQRCode::QRCode.new qrcode_content, level: :l, size: 2
+          alignment = page_num.even? ? :right : :left
+          render_qr_code(qrcode, align: alignment, dot: 4.0, stroke: false)
+          draw_text(qrcode_content,
+                    at: [alignment == :left ? 140 : bounds.width - 140 - qrcode_content.length * 6, bounds.height - 30])
+          start_new_page
         end
-        combine_pdf_qr = CombinePDF.parse(pdf.render)
-        template_pdf.pages.zip(combine_pdf_qr.pages) do |template_page, qr_page|
-          generated_pdf << (qr_page << template_page)
-        end
-        progress.increment
       end
-
-      generated_pdf.save File.join(
-        exam_template.base_path,
-        "#{exam_template.name}-#{start}-#{start + num_copies - 1}.pdf"
-      )
-     end
-     m_logger.log('Generate pdf copies process done')
-    rescue => e
-      Rails.logger.error e.message
-      raise e
+      combine_pdf_qr = CombinePDF.parse(pdf.render)
+      template_pdf.pages.zip(combine_pdf_qr.pages) do |template_page, qr_page|
+        generated_pdf << (qr_page << template_page)
+      end
+      progress.increment
+    rescue StandardError => e
+      status.update(error_message: e.to_s)
+      raise
```

Comment:
I might be missing something here, but is this `raise` actually going to be caught anywhere? The pattern seems different than all the other jobs, where there's a top-level `rescue` that doesn't raise an error.

Reply:
There is a `rescue_from` in active job status that should catch all remaining raised errors. The reason this re-raises the error is actually to appease rubocop which doesn't like plain return statements from within a loop. Re-raising the error seemed to be the cleanest way to exit the job early while keeping rubocop happy. 

Alternatives are to return here or to break and then to include some logic afterwards to prevent an incomplete pdf from being created.

Reply:
this has been resolved because error catching has been moved to the `rescue_from` block in `ApplicationJob`